src/sql/SqlService.ts

```diff
@@ -0,0 +1,290 @@
+            return value.toString();
+        } else if (value instanceof HzOffsetDateTime) {
+            return value.toISOString();
+        } else {
+            return value;
+        }
+    }
+
+    execute(sql: SqlStatement): SqlResult;
+    execute(sql: string, params?: any[], options?: SqlStatementOptions): SqlResult;
+    execute(sql: string | SqlStatement, params?: any[], options?: SqlStatementOptions): SqlResult {
+        let sqlStatement: SqlStatement;
+
+        if (typeof sql === 'string') {
+            sqlStatement = {
+                sql: sql
+            };
+            if (params !== undefined) { // todo: document this
+                sqlStatement.params = params;
+            }
+            if (options !== undefined) {
+                sqlStatement.options = options;
+            }
+        } else if (typeof sql === 'object') { // assume SqlStatement is passed
+            sqlStatement = sql;
+        } else {
+            throw new IllegalArgumentError('Sql can be an object or string');
+        }
+
+        try {
+            SqlServiceImpl.validateSqlStatement(sqlStatement);
+        } catch (error) {
+            throw new IllegalArgumentError(`Invalid argument given to execute(): ${error.message}`, error)
+        }
+
+        const connection = this.connectionRegistry.getRandomConnection(true);
+        if (connection === null) {
+            throw new HazelcastSqlException(
+                this.connectionManager.getClientUuid(),
+                SqlErrorCode.CONNECTION_PROBLEM,
+                'Client is not currently connected to the cluster.'
+            );
+        }
+
+        const queryId = SqlQueryId.fromMemberId(connection.getRemoteUuid());
+
+        const expectedResultType: SqlExpectedResultType =
+            sqlStatement.options?.expectedResultType ? SqlExpectedResultType[sqlStatement.options.expectedResultType]
+                : SqlServiceImpl.DEFAULT_EXPECTED_RESULT_TYPE;
+
+        try {
+            const serializedParams = [];
+            if (Array.isArray(sqlStatement.params)) { // params can be undefined
+                for (let param of sqlStatement.params) {
+                    param = this.convertToStringIfDatetimeValue(param);
+                    serializedParams.push(this.serializationService.toData(param));
+                }
+            }
+            const cursorBufferSize = sqlStatement.options?.cursorBufferSize ?
+                sqlStatement.options.cursorBufferSize : SqlServiceImpl.DEFAULT_CURSOR_BUFFER_SIZE;
+            const requestMessage = SqlExecuteCodec.encodeRequest(
+                sqlStatement.sql,
+                serializedParams,
+                sqlStatement.options?.timeoutMillis ? sqlStatement.options.timeoutMillis : SqlServiceImpl.DEFAULT_TIMEOUT,
+                cursorBufferSize,
+                sqlStatement.options?.schema ? sqlStatement.options.schema : SqlServiceImpl.DEFAULT_SCHEMA,
```

Comment:
what if the `schema` is an empty string (or any falsy string value)

Reply:
`(or any falsy string value)` which is empty string. Handled empty string thanks. FYI, `sqlStatement.options?.schema` returns undefined if options is null or undefined.